src/state-calculator.tsx

```diff
@@ -432,16 +298,159 @@ export const StateCalculator: FC<{
           />
         </>
       )}
-    </>
+    </div>
   );
 };
 
+class CalculatorElement extends HTMLElement {
+  /* supported property to switch UI language */
+  override lang: string = this.getDefaultLanguage();
+
+  /* property to show the email signup field */
+  showEmail: boolean = false;
+
+  /* property to include incentives from states that aren't formally launched */
+  includeBetaStates: boolean = false;
+
+  /* supported properties to control which API path and key is used to load the calculator results */
+  apiKey: string = '';
+  apiHost: string = DEFAULT_CALCULATOR_API_HOST;
+
+  /* supported property to allow restricting the calculator to ZIPs in a specific state */
+  state: string = '';
+
+  /* supported properties to allow pre-filling the form */
+  zip: string = DEFAULT_ZIP;
+  ownerStatus: OwnerStatus = DEFAULT_OWNER_STATUS;
+  householdIncome: string = DEFAULT_HOUSEHOLD_INCOME;
+  taxFiling: FilingStatus = DEFAULT_TAX_FILING;
+  householdSize: string = DEFAULT_HOUSEHOLD_SIZE;
+
+  /* attributeChangedCallback() will be called when any of these changes */
+  static observedAttributes = [
+    'lang',
+    'show-email',
+    'include-beta-states',
+    'api-key',
+    'api-host',
+    'state',
+    'zip',
+    'owner-status',
+    'household-income',
+    'tax-filing',
+    'household-size',
+  ] as const;
+
+  reactRootCalculator: Root;
+  reactRootFooter: Root;
+
+  constructor() {
+    super();
+
+    const shadowRoot = this.attachShadow({ mode: 'open' });
+
+    const style = document.createElement('style');
+    style.textContent = tailwindStyles;
+
+    shadowRoot.appendChild(style);
+
+    const calculator = document.createElement('div');
+    shadowRoot.appendChild(calculator);
+    this.reactRootCalculator = createRoot(calculator);
+
+    const footer = document.createElement('div');
+    shadowRoot.appendChild(footer);
+    this.reactRootFooter = createRoot(footer);
+  }
+
+  connectedCallback() {
+    this.render();
+  }
+
+  attributeChangedCallback(
+    attr: (typeof CalculatorElement.observedAttributes)[number],
+    _: string | null, // old value; unused
+    newValue: string | null,
+  ) {
+    if (attr === 'lang') {
```

Comment:
Does Typescript help us if we miss an attr here? Would a switch statement help?

Idle thought, not a blocker. Wall of `else if` or wall of `case` - the difference seems superficial if there's no tooling/linting benefit.

Reply:
Yes, see line 399. The type of `attr` is a big union of constants, and each `else if` case removes one constant from it. In the final `else`, it should be an empty union, i.e. `never`, and assigning it to a variable of type `never` asserts that.

A switch could do the same, but I did it this way to avoid the noise of `break;` in every case.